MeasureGui: Revert temporary measure object creation

hlorus opened the pull request:
As discussed in #16815, this reverts changes made in #15122 such that measurements are created with the document's addObject method.

This reopens #13735

wwmayer reviewed (changes requested):
`_mDocument` and `_mGuiDocument` are private members of  `TaskMeasure`  that are not used any more.  These members must be removed to avoid possible compiler warnings.

wwmayer commented on src/Mod/Measure/Gui/TaskMeasure.cpp:
```diff
@@ -174,85 +174,36 @@ void TaskMeasure::enableAnnotateButton(bool state)
 
+        // Create a MeasurePython instance
+        auto featurePython = doc->addObject("Measure::MeasurePython", measureType->label.c_str());
+        _mMeasureObject = (Measure::MeasureBase*)featurePython;
```
Do not use C-style casts. They are considered harmful because they can act like `static_cast`, `reinterpret_cast` or `const_cast` or a combination of them and you don't know which of the casts are actually performed.

Use a `dynamic_cast` or `static_cast` here.

wwmayer commented on src/Mod/Measure/Gui/TaskMeasure.cpp:
```diff
@@ -174,85 +174,36 @@ void TaskMeasure::enableAnnotateButton(bool state)
-        _mViewObject = nullptr;
+    else {
+        // Create measure object
+        _mMeasureObject = (Measure::MeasureBase*)doc->addObject(measureType->measureObject.c_str(),
```
Same here.

wwmayer commented on src/Mod/Measure/Gui/TaskMeasure.h:
```diff
@@ -65,15 +65,13 @@ class TaskMeasure: public TaskView::TaskDialog, public Gui::SelectionObserver
     bool hasSelection();
     void clearSelection();
     bool eventFilter(QObject* obj, QEvent* event) override;
-    void setMeasureObject(Measure::MeasureBase* obj);
```
TaskMeasure.h lacks of a file guard. Please add one as otherwise compilation can fail due to duplicated class declarations.

wwmayer commented:
Some extra issues that are not part of this PR but can cause problems:
* TaskMeasure installs itself as event filter and overrides eventFilter() to handle Esc, Return/Enter key events. This can lead to misbehaviour of other widgets that have the focus. Is the event filter really necessary?
* canAnnotate is a global function but neither is declared as static nor it is part of anonymous namespace. This can cause undefined behaviour that leads to linking errors or runtime crashes.
* TaskMeasure::update() can cause a crash because of the assert()
1. Create two documents, each with an object
2. Start the measurement and click on the object
3. Switch to the other document and click on the object
4. `assert(viewObject);` fails and terminates the application
* `TaskMeasure::ensureGroup` can cause undefined behaviour:
1. Create a document
2. Run this in the Python console
```
App.ActiveDocument.addObject("Part::Box", "Measurements")
App.ActiveDocument.recompute()
```
3. Start the measurement and click on the box
4. Click on Save.
5. Crash! This is because of the blind `static_cast` of the accessed object
* getStandardButtons() defines the buttons Apply, Abort, Reset and modifyStandardButtons() changes the text of the Abort button to "Close". So, why doesn't use getStandardButtons() the Close button instead?
* The same with the Apply button that is relabeled to "Save"

hlorus commented:
I've addressed all comments except for the following:

> Some extra issues that are not part of this PR but can cause problems:
> * TaskMeasure installs itself as event filter and overrides eventFilter() to handle Esc, Return/Enter key events. This can lead to misbehaviour of other widgets that have the focus. Is the event filter really necessary?

Without the event filter the task immediatley looses focus as soon as an element is selected, therefor esc/enter won't get handled by the task. This makes the shortcuts basically unusable.


> * getStandardButtons() defines the buttons Apply, Abort, Reset and modifyStandardButtons() changes the text of the Abort button to "Close". So, why doesn't use getStandardButtons() the Close button instead?
> * The same with the Apply button that is relabeled to "Save"

Didn't have time for that, sorry. I assume this can also be addressed in a separate PR.

wwmayer commented:
> Without the event filter the task immediatley looses focus as soon as an element is selected, therefor esc/enter won't get handled by the task. This makes the shortcuts basically unusable.

But that's the case for basically all task dialogs and I don't consider this a big issue.

chennes commented:
@wwmayer if @hlorus has addressed all of your required changes (I see that there are some additional suggestions that it seems can be part of a later PR), could you mark those changes as complete in GitHub?

wwmayer commented:
Everything but two issues are fixed now:
* The Abort/Apply/Reset buttons are used and relabeled instead of Close/Save/Reset
* By using the event filter the measure dialog can steal events of even a modal dialog that can lead to weird behaviour

But both can be addressed in dedicated PRs.

yorikvanhavre commented:
backported
